scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -436,112 +444,237 @@
+                (log/debug "Matched tasks" task-txns)
+                ;; This launch-tasks MUST happen after the above transaction in
+                ;; order to allow a transaction failure (due to failed preconditions)
+                ;; to block the launch
+                (meters/mark! scheduler-offer-matched
+                              (->> matches
+                                   (mapcat (comp :id :offer :leases))
+                                   (distinct)
+                                   (count)))
+                (meters/mark! matched-tasks (count task-txns))
+                (meters/mark! matched-tasks-cpus (:cpus match-resource-requirements))
+                (meters/mark! matched-tasks-mem (:mem match-resource-requirements))
+                (doseq [{:keys [tasks leases]} matches
+                        :let [offers (mapv :offer leases)
+                              task-infos (mapv (fn process-results [^TaskAssignmentResult task]
+                                                 (reduce
+                                                   (fn add-ports-to-task-info [task-info [index port]]
+                                                     (log/debug "task-info" task-info [index port])
+                                                     (-> task-info
+                                                         (update-in [:resources :ports]
+                                                                    (fnil conj [])
+                                                                    {:begin port :end port})
+                                                         (assoc-in [:command :environment (str "PORT" index)]
+                                                                   (str port))))
+                                                   (:task-info (.getRequest task))
+                                                   (map-indexed (fn [index port] [index port])
+                                                                (.getAssignedPorts task))))
+                                               tasks)
+                              slave-id (:slave-id (first offers))]]
+                  (mesos/launch-tasks
+                    driver
+                    (mapv :id offers)
+                    (mapv #(-> %
+                               (assoc :slave-id slave-id)
+                               (dissoc :num-ports))
+                          task-infos))
+                  (doseq [^TaskAssignmentResult task tasks]
+                    (.. fenzo
+                        (getTaskAssigner)
+                        (call (.getRequest task) (get-in (first leases) [:offer :hostname])))))
+                (:matched-head? processed-matches)))))
       (catch Throwable t
-        (decline-offers driver (:ids offer))
         (meters/mark! handle-resource-offer!-errors)
-        (log/error t "Error in match:" (ex-data t))))))
+        (log/error t "Error in match:" (ex-data t))
```

Comment:
The difference in return value is used to determine whether to lower the number of jobs Fenzo can look at, right?  It's not clear to me why we want to make sure that if handle-resource-offer! errors out, we want to "penalize" Fenzo.  The previous version of this code did penalize Fenzo for this by returning something falsey though.   Do you know more about the reasoning here?